graphql/resolvers/catalog/Product.ts

```diff
@@ -0,0 +1,58 @@
+import {juxt, map, omit} from 'ramda'
+import {objToNameValue} from './objToNameValue'
+import {resolveBuy, resolveView} from './recommendationsResolver'
+
+const knownNotPG = [
+  'allSpecifications',
+  'brand',
+  'categories',
+  'categoriesIds',
+  'categoryId',
+  'clusterHighlights',
+  'description',
+  'items',
+  'productId',
+  'productName',
+  'link',
+  'linkText',
+  'productReference',
+]
+
+export default {
+  Product: {
+    clusterHighlights: (product) => {
+      const {clusterHighlights={}} = product
+      return objToNameValue('id', 'name', clusterHighlights)
+    },
+
+    propertyGroups: (product) => {
+      const {allSpecifications=[]} = product
+      const notPG = knownNotPG.concat(allSpecifications)
+      return objToNameValue('name', 'properties', omit(notPG, product))
+    },
+
+    properties: (product) => {
+      const {allSpecifications=[]} = product
+      return map(name => ({name, values: product[name]}), allSpecifications)
+    },
+
+    recommendations: async (product, args, {vtex: ioContext}) => {
+      const [view, buy] = await Promise.all([
+        resolveView(ioContext, product),
+        resolveBuy(ioContext, product)
+      ])
+      return {buy, view}
+    }
+  },
+
+  SKU: {
+    variations: (sku) => {
```

Comment:
How does this work? Who passes the variable `sku`? Why it doesn't have to fetch anything?

Reply:
I know, looks like magic, right? ✨ 

This resolver is only ever called when an object of the type `SKU` has already been returned somewhere, such as in a query. The fetched object is then automatically passed to a resolver for each of the type's fields as the first parameter, as can be seen in this resolver for the `variations` field. So, basically, the first parameter of any resolver will be the `root` object for that type (or `null` if the type is `Query` or `Mutation`, since those are the only places that don't have information yet, where the request starts.

In this case, `SKU` is only ever used inside the `Product` type, which is only returned by the `product` and `products` queries. Products are always fetched from the same url, which returns a whole object that includes all skus with their `variations`. So, when this resolver is called, GraphQL automatically gives it everything it knows about the `SKU` object that had already been fetched and we can be sure of what the field resolver will receive.

This can be really useful, specially in types that are used by many different places, to ensure data normalization and special treatments. Imagine that products are returned in 10 different queries for example. By having a resolver for the `recommendations` field you can guarantee that all these places are covered. The only drawback is that you have to know which fields from the root object will always be available. If you need the product's `price` in a resolver, but that's not always fetched from the server and you try to use it, it would crash. But that's not usually a problem.

Sorry for the long text, just wanted to try to explain in details for everyone. But if you find this confusing you can also try the official GraphQL docs, which also explain this since it's the pattern for GraphQL servers: http://graphql.org/learn/execution/#root-fields-resolvers.